src/host/ft_host/API_MultipleInflightMessageTests.cpp

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+
+class MultipleInflightMessageTests
+{
+    BEGIN_TEST_CLASS(MultipleInflightMessageTests)
+    END_TEST_CLASS()
+
+    TEST_METHOD(WriteWhileReadingInputCrash)
+    {
+        BEGIN_TEST_METHOD_PROPERTIES()
+            TEST_METHOD_PROPERTY(L"IsolationLevel", L"Method") // Don't pollute other tests by isolating our potential crash and buffer resizing to this test.
+        END_TEST_METHOD_PROPERTIES()
+
+        using namespace std::string_view_literals;
+
+        const auto inputHandle = GetStdInputHandle();
+        const auto outputHandle = GetStdOutputHandle();
+
+        DWORD originalConsoleMode{};
+        VERIFY_WIN32_BOOL_SUCCEEDED(GetConsoleMode(inputHandle, &originalConsoleMode));
+        auto restoreMode{ wil::scope_exit([=]() {
+            SetConsoleMode(inputHandle, originalConsoleMode);
+        }) };
+        VERIFY_WIN32_BOOL_SUCCEEDED(SetConsoleMode(inputHandle, ENABLE_ECHO_INPUT | ENABLE_LINE_INPUT));
+
+        // Prime the console with some fake input records.
+        const std::array<INPUT_RECORD, 4> inputRecords{
+            INPUT_RECORD{ KEY_EVENT, KEY_EVENT_RECORD{ TRUE, 1, 'h', 0, L'h', 0 } },
+            INPUT_RECORD{ KEY_EVENT, KEY_EVENT_RECORD{ FALSE, 1, 'h', 0, L'h', 0 } },
+            INPUT_RECORD{ KEY_EVENT, KEY_EVENT_RECORD{ TRUE, 1, 'i', 0, L'i', 0 } },
+            INPUT_RECORD{ KEY_EVENT, KEY_EVENT_RECORD{ FALSE, 1, 'i', 0, L'i', 0 } },
+        };
+        DWORD written{};
+        VERIFY_WIN32_BOOL_SUCCEEDED(WriteConsoleInput(inputHandle, inputRecords.data(), static_cast<DWORD>(inputRecords.size()), &written));
+        VERIFY_ARE_EQUAL(inputRecords.size(), written);
+
+        // !!! LOAD BEARING !!!
+        // This buffer needs to be large enough to force API_MSG to heap allocate (!)
+        std::array<wchar_t, 129> buffer;
+        DWORD read{};
+
+        bool readerThreadLaunched{ false };
+        std::condition_variable readerThreadLaunchCV;

```

Comment:
FYI: You can build yourself a super simple barrier:
```cpp
struct latch
{
    explicit latch(int slots)
    {
        InitializeSynchronizationBarrier(&_barrier, slots, -1);
    }

    ~latch()
    {
        DeleteSynchronizationBarrier(&_barrier);
    }

    void arrive_and_wait()
    {
        EnterSynchronizationBarrier(&_barrier, 0);
    }

private:
    SYNCHRONIZATION_BARRIER _barrier{};
};
```

I also uploaded a full std::latch, that I built a while ago: https://gist.github.com/lhecker/f12f9e9fa3a60ab71f44314c7df74da6

Reply:
oh, i like that. clever. :)